integration-tests/environments/react-native/metro.config.js

```diff
@@ -16,40 +16,96 @@
 //
 ////////////////////////////////////////////////////////////////////////////
 
+// Metro bundler does not support symlinks (see https://github.com/facebook/metro/issues/1).
+// In order to link to our local version of Realm JS in a way which Metro can follow,
+// we need to instead add the module to `resolver.blockList`, and then in `resolver.extraNodeModules`,
+// override the locations for any symlinked modules by specifying a mapping from their name
+// to their actual location on disk.
+
+// In order to achieve this, we recursively check if each file/directory in `node_modules`
+// is a symlink or not. If it is, and if it is specified in our main `package.json` (i.e.
+// it is a direct dependency of our app), we add it to the `blockList` and
+// add a mapping for it. We also add it to the list of watched folders, so
+// that any changes to it are picked up by hot reloading.
+
 const path = require("path");
 const fs = require("fs");
 const exclusionList = require("metro-config/src/defaults/exclusionList");
 
-// Read the localDependencies property of package.json and resolve the relative path values
-const appPackageJson = readJson(__dirname, "package.json");
-const localDependencies = Object.fromEntries(
-  Object.entries(appPackageJson.localDependencies || {}).map(([localName, relativePath]) => [
-    localName,
-    path.resolve(__dirname, relativePath),
-  ]),
-);
-
 function readJson(...pathSegemnts) {
   const filePath = path.join(...pathSegemnts);
   return JSON.parse(fs.readFileSync(filePath, "utf8"));
 }
 
-const watchFolders = Object.values(localDependencies);
+function getLinkedDependencies(packagePath, recursive = true, exclude = new Set()) {
+  const packageJson = readJson(packagePath, "package.json");
+  const nodeModulesPath = path.join(packagePath, "node_modules");
+  const files = fs.readdirSync(nodeModulesPath, { encoding: "utf8", withFileTypes: true });
+  const links = files
+    .filter((f) => f.isSymbolicLink())
+    .map(({ name }) => {
+      const linkPath = fs.realpathSync(path.join(nodeModulesPath, name));
+      const linkPackageJson = readJson(linkPath, "package.json");
+      const peerDependencies = Object.keys(linkPackageJson.peerDependencies || {});
+      return { name, path: linkPath, peerDependencies };
+    });
+  // We're only interested in actual dependencies
+  const dependencyLinks = links.filter(({ name }) => name in packageJson.dependencies && !exclude.has(name));
```

Comment:
In `realm-react`, I have some conflicts with `devDependencies` as well, since the tests make use of the `babel-runner` and `react-native`. Might be worthwhile to check `name in packageJson.devDependencies` as well.